docs/proposals/20230612-node-os-image-caching.md

```diff
@@ -0,0 +1,268 @@
+
+Example AzureMachinePoolMachine yaml with the important fields shown (want status.ready == true, latestModelApplied: true and each status: "True"):
+
+```yaml
+apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
+kind: AzureMachinePool
+metadata:
+  name: node-os-image-caching-machine-pool-machine
+  namespace: default
+  creationTimestamp: "2023-06-20T17:41:54Z"
+status:
+  conditions:
+  - lastTransitionTime: "2023-06-20T17:43:39Z"
+    status: "True"
+    type: Ready
+  - lastTransitionTime: "2023-06-20T17:43:39Z"
+    status: "True"
+    type: BootstrapSucceeded
+  - lastTransitionTime: "2023-06-20T17:43:39Z"
+    status: "True"
+    type: NodeHealthy
+  latestModelApplied: true
+  ready: true
+```
+
+In terms of when to take a snapshot, a day is given as a general example which should be good for typical use but the specification of how often will be customizable as we know that certain operators have different strategies and use cases for how they’re running their services on our clusters.
+
+In terms of data model changes, AzureMachinePool will be changed and the changes we expect will be purely additive and nonbreaking. No removals should be required to the data model. For AzureMachinePool we will add a new optional field under spec.template.image called nodePrototyping which will be enabled if present and it have a required field under it called interval which will map to an interval of 1 day or 24 hours by default.
+
+For AzureMachinePool, we will also add an optional field under spec.strategy.rollingUpdate called prototypeAutomaticRollout which will be set to false by default since we don't want the replacement of a node OS image to automatically trigger a rolling update of all the nodes. This is because the new image will be functionally identical to the old one outside of name itself (since the old nodes will all have the update and security patch contents already present). If for some reason the operator wants to always trigger this rollout (maybe if they programmatically use the image names themselves), then they can simply set this field to true.
```

Comment:
I am still relatively confident that we can't do it with the existing rollingUpdate parameters because of the fact that normal rollingUpdate still needs to work as the user intended, so we can't really hyjack any of those fields

Reply:
@CecileRobertMichon @mboersma Do either of you have a good idea for where to put this field? Optionally under rolling update is awkward and it can't be its own strategy because it needs to be usable with someone using rollingupdate for other general things

Reply:
As discussed offline, this doesn't need to be in scope for the initial proposal. I agree with @nojnhuh we should try to reuse what's there already as having a second "rollingUpdate" field in the API makes the API ambiguous and redundant.

See also https://cluster-api.sigs.k8s.io/tasks/upgrading-clusters.html?highlight=rolloutA#how-to-schedule-a-machine-rollout

Reply:
Ended up removing this out of the proposal entirely because I think at this point it's both out of scope/any other uses the operator has can be covered by scheduling an after rollout.